plugin/documents.py

```diff
@@ -136,9 +136,17 @@ def detach(self) -> None:
         self.ids_to_listeners.pop(self.buffer.buffer_id, None)
         super().detach()
 
-    def on_text_changed_async(self, changes: Iterable[sublime.TextChange]) -> None:
-        for listener in list(self.view_listeners):
-            listener.on_text_changed_async(changes)
+    def on_text_changed(self, changes: Iterable[sublime.TextChange]) -> None:
+        view = self.buffer.primary_view()
+        if not view:
+            return
+        change_count = view.change_count()
+
+        def notify() -> None:
+            for listener in list(self.view_listeners):
+                listener.on_text_changed_async(change_count, changes)
+
+        sublime.set_timeout_async(notify)
```

Comment:
Will try to see if the timeout is really necessary.

Reply:
Yeah, it works even without the timeout:
```diff
         change_count = view.change_count()
 
-        def notify() -> None:
-            for listener in list(self.view_listeners):
-                listener.on_text_changed_async(change_count, changes)
-
-        sublime.set_timeout_async(notify)
+        for listener in list(self.view_listeners):
+            listener.on_text_changed_async(change_count, changes)
```

Reply:
That would run a few `Session.send_request` calls in the main thread. The idea is to move all calls to the worker thread instead, and then remove that one lock in the Session class. All the response handlers and notification handlers already run in the worker thread.